--- fuse-0.8.0.1/ui/fb/fbkeyboard.c.old	2007-06-17 12:14:29.000000000 +0200
+++ fuse-0.8.0.1/ui/fb/fbkeyboard.c	2007-06-17 12:17:00.000000000 +0200
@@ -134,12 +134,16 @@
       } else if( ( keybuf[i] & 0x7f ) == 0 ) {
 	ignore = 2; /* ignore extended keysyms */
       } else {
+      	input_key fuse_keysym;
 	input_event_t fuse_event;
 
 	fuse_event.type = ( keybuf[i] & 0x80 ) ?
                           INPUT_EVENT_KEYRELEASE :
                           INPUT_EVENT_KEYPRESS;
-	fuse_event.types.key.native_key = keysyms_remap( keybuf[i] & 0x7f );
+
+        fuse_keysym = keysyms_remap( keybuf[i] & 0x7f );
+	fuse_event.types.key.native_key = fuse_keysym;
+	fuse_event.types.key.spectrum_key = fuse_keysym;
 
 	if( fuse_event.types.key.native_key ) input_event( &fuse_event );
       }
--- fuse-0.8.0.1/ui/svga/svgakeyboard.c.old	2007-06-17 12:12:20.000000000 +0200
+++ fuse-0.8.0.1/ui/svga/svgakeyboard.c	2007-06-17 12:13:29.000000000 +0200
@@ -76,6 +76,7 @@
 
   fuse_event.type = INPUT_EVENT_KEYPRESS;
   fuse_event.types.key.native_key = fuse_keysym;
+  fuse_event.types.key.spectrum_key = fuse_keysym;
 
   return input_event( &fuse_event );
 }
@@ -92,6 +93,7 @@
 
   fuse_event.type = INPUT_EVENT_KEYRELEASE;
   fuse_event.types.key.native_key = fuse_keysym;
+  fuse_event.types.key.spectrum_key = fuse_keysym;
 
   return input_event( &fuse_event );
 }
